File: driver/bridge/src/capabilities.rs
//! Driver capability handshake — NOT a rkyv wire type.
//!
//! Each driver advertises its capabilities (page geometry, forward limits,
//! tokenizer arch, etc.) at startup via a JSON blob delivered to the runtime.
//! The runtime parses it into [`DriverCapabilities`] for scheduling and
//! batching decisions. The shape is wire-stable — fields appear verbatim
//! in the driver's ready-callback JSON (see `driver/cuda/src/entry.cpp`,
//! `driver/portable/src/entry.cpp`, `driver/dummy/src/lib.rs`).
//!
//! This is intentionally not under `#[schema]` — the handshake happens
//! once at driver startup and uses JSON over a side channel, not the
//! rkyv ring.

use serde::{Deserialize, Serialize};

/// Static driver capabilities reported at handshake time.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DriverCapabilities {
    /// Total KV pages available for context residency.
    pub total_pages: u32,
    /// KV page size in tokens.
    pub kv_page_size: u32,
    /// Number of CPU-resident swap-pool pages (0 if no swap support).
    pub swap_pool_size: u32,
    /// Maximum forward-pass tokens accepted in one driver fire.
    pub max_forward_tokens: u32,
    /// Maximum forward-pass requests accepted in one driver fire.
    pub max_forward_requests: u32,
    /// Maximum page references accepted in one driver fire.
    pub max_page_refs: u32,
    /// Maximum logits rows the driver can return in one fire.
    pub max_logit_rows: u32,
    /// Maximum probability rows the driver can return in one fire.
    pub max_prob_rows: u32,
    /// Maximum custom-mask bytes accepted in one fire.
    pub max_custom_mask_bytes: u32,
    /// Maximum sampler rows accepted in one fire.
    pub max_sampler_rows: u32,
    /// Maximum logprob label count accepted in one fire.
    pub max_logprob_labels: u32,
    /// Architecture name (e.g. `llama3`, `qwen3`) — used for tokenizer dispatch.
    pub arch_name: String,
    /// Vocabulary size — pinned by the loaded model.
    pub vocab_size: u32,
    /// Maximum model context length (positions). Drives scheduler ceiling.
    pub max_model_len: u32,
    /// Activation dtype on the driver side (`bf16` / `f16` / `f32`).
    pub activation_dtype: String,
    /// Optional snapshot directory the driver can use to persist state.
    #[serde(default)]
    pub snapshot_dir: String,
    /// Optional shmem region name; only Some when the driver speaks shmem
    /// (subprocess flavors). In-process drivers leave this absent.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub shmem_name: Option<String>,
}

File: driver/dev/src/pie_driver_dev/_bridge/capabilities.py
"""Driver → runtime handshake contract.

Pie's configuration has two distinct concepts that this module separates:

  * **Budgets** (top-down, user-set in `ModelConfig` / `SchedulerConfig`):
    "you may use up to this much" — `gpu_mem_utilization`, `swap_budget_bytes`,
    `default_token_limit`, scheduler timing knobs, etc. These flow from the
    user's config file into the worker and are consumed there. They never
    come back.

  * **Capacities** (bottom-up, driver-computed in this dataclass): "given
    that budget plus the kernel's real ceiling, here is what I actually
    have" — number of KV pages, page size the attention impl supports,
    forward limits, model-architecture facts read from HF config. These
    flow from the worker back to the server via the `ready_queue` handshake
    and are then forwarded to the Rust runtime when applicable.

Each inference driver (`pie_driver`, `pie_driver_vllm`, future siblings)
implements `Engine.capabilities() -> DriverCapabilities` and the worker
delivers that struct as the ready signal payload. The server uses it as the
source of truth when constructing `pie_runtime.ModelConfig`.

Some fields are forwarded to Rust through existing `pie_runtime.ModelConfig` /
`DeviceConfig` kwargs; others are Python-only for now (used for validation
and logging at the server layer). See server._bootstrap for the wiring.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class DriverCapabilities:
    """Capacities the inference driver reports to pie's runtime.

    Frozen so the value can't drift after the handshake — once the worker
    publishes it, both Python and Rust treat it as immutable truth for the
    lifetime of that worker.
    """

    # ── KV cache ────────────────────────────────────────────────────────
    # Number of KV pages the driver allocated on this rank's GPU.
    # Forwarded to Rust as DeviceConfig.total_pages.
    total_pages: int

    # Block size in tokens that the chosen attention kernel actually uses.
    # Resolved by the driver and forwarded to Rust as ModelConfig.kv_page_size.
    kv_page_size: int

    # Pinned host-side KV slot count for swap-out. 0 = swap disabled.
    # Forwarded to Rust as DeviceConfig.cpu_pages.
    swap_pool_size: int

    # ── Forward limits ─────────────────────────────────────────────────
    # Clamped to what the driver's kernel actually supports. Dynamic
    # allocation drivers can use u32::MAX for sub-limits that do not apply.
    max_forward_tokens: int
    max_forward_requests: int
    max_page_refs: int
    max_logit_rows: int
    max_prob_rows: int
    max_custom_mask_bytes: int
    max_sampler_rows: int
    max_logprob_labels: int

    # ── Model-architecture facts ───────────────────────────────────────
    # First architecture string from the HF config (e.g. "LlamaForCausalLM",
    # "Qwen3ForCausalLM"). Forwarded to Rust as ModelConfig.arch_name.
    arch_name: str

    # Vocabulary size and max context length. Currently Python-only —
    # surfaced for telemetry and future Rust-side admission control.
    vocab_size: int
    max_model_len: int

    # Resolved activation dtype, never "auto". Currently Python-only.
    activation_dtype: str

    # ── Feature support ────────────────────────────────────────────────
    # Whether the driver implements user-supplied attention masks (i.e.,
    # respects `req.has_user_mask` from pie's runtime). False means the
    # driver computes plain causal attention regardless of any inferlet-
    # supplied mask. Inferlets that depend on non-causal patterns must be
    # routed to a driver that returns True (currently only `native`).
    # Defaults False so a driver that omits the field is treated as
    # mask-unaware. Python-only today; the Rust runtime gains an admission
    # hook in a follow-up.
    supports_user_attention_mask: bool = False

    # Whether the driver implements CMA-ES / LoRA adapter operations
    # (`init_adapter`, `update_adapter`, `load_adapter`, `save_adapter`,
    # plus per-batch noise injection at Q/K/V). False means the driver
    # raises NotImplementedError on adapter calls; inferlets that exercise
    # adapter paths must run on a driver that returns True (currently only
    # `native`). Defaults False so a driver that omits the field is treated
    # as adapter-unaware. Python-only today; the Rust runtime gains an
    # admission hook in a follow-up.
    supports_adapters: bool = False

    # ── Filesystem ─────────────────────────────────────────────────────
    # Local directory containing tokenizer.json + config.json. The Rust
    # runtime reads tokenizer.json from `<snapshot_dir>/tokenizer.json`.
    # Forwarded to Rust as part of ModelConfig.tokenizer_path.
    snapshot_dir: str = ""
